src/installer.ts

```diff
@@ -161,7 +161,10 @@ async function apt(tool: Tool, version: string): Promise<void> {
   const toolName = tool === 'ghc' ? 'ghc' : 'cabal-install';
   const v = tool === 'cabal' ? version.slice(0, 3) : version;
   core.info(`Attempting to install ${toolName} ${v} using apt-get`);
-  await exec(`sudo -- sh -c "apt-get -y install ${toolName}-${v}"`);
+  // Ignore the return code so we can fall back to ghcup
+  await exec(`sudo -- sh -c "apt-get -y install ${toolName}-${v}"`, undefined, {
```

Comment:
While this is a valid approach, it only ignores the errors for apt-get. I had originally designed these functions to be used through the installTool function which should ideally fall back completely through to all methods. So if the apt function is not falling back correctly, I don't think any of them are; if not, this logic should be introduced to the `stack`, `ghcup`, and `choco` installation functions as well. (Despite stack not having a fallback, it'd be nice if the correct failed message showed up)

Ideally, a test would be introduced to prevent a regression on the desired fallback behavior. There's not a good way to test it without adding something into the CI, but attempting to install a non-existing version of ghc, cabal, and stack should force all of the fallbacks to happen correctly.

Reply:
I'm not sure if I would know how to go about setting up the test for that, would that need to be done in this PR?

Reply:
I don't think so. The way I usually setup that sort of test is to just add something to the `.github/workflows/workflow.yml` file.

Pretty sure this would do what we want (but I haven't tried it yet)

```yaml
  test-haskell-fallbacks:
    name: GHC ${{ matrix.ghc }}, Cabal ${{ matrix.cabal }} - ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    continue_on_error: true # this test should fail
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, macOS-latest, windows-latest]
        ghc: ['90.90.90']
        cabal: ['90.90']
    steps:
      - uses: actions/checkout@v2
      - uses: ./
        with:
          ghc-version: ${{ matrix.ghc }}
          cabal-version: ${{ matrix.cabal }}
```

Reply:
@bubba can you provide a little more context on what prompted needing to ignore return codes here. What wasn't working?